File: fenstrings/fenparser.hs
-- fenparser.hs
-- Run with:
-- $ stack runghc fenparser.hs "input goes here"

module FENParser (main) where

import Control.Monad
import Data.Char
import Data.Void
import System.Environment

import Text.Megaparsec
import Text.Megaparsec.Char
import qualified Text.Megaparsec.Char.Lexer as L

type Parser = Parsec Void String 

data PieceType
  = Rook
  | Knight
  | Bishop
  | Queen
  | King
  | Pawn
  deriving (Eq, Show)

data TeamColor
  = TeamColorWhite
  | TeamColorBlack
  deriving (Eq, Show)

type Piece = (PieceType, TeamColor)

-- Can be a piece or a 'skip' digit
data PiecePlacement
  = PlacementPiece Piece
  | PlacementInt Int
  deriving (Eq, Show)

data CastlingAvailability
  = WhiteKingside
  | WhiteQueenside
  | BlackKingside
  | BlackQueenside
  deriving (Eq, Show)

data Rank
  = RankOne
  | RankTwo
  | RankThree
  | RankFour
  | RankFive
  | RankSix
  | RankSeven
  | RankEight
  deriving (Eq, Show)

data File
  = FileA
  | FileB
  | FileC
  | FileD
  | FileE
  | FileF
  | FileG
  | FileH
  deriving (Eq, Show)

-- Piece placements
-- rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
-- ^-----------------------------------------^

piece :: Parser Piece
piece = choice [
  (Pawn,    TeamColorWhite) <$ char 'P',
  (Knight,  TeamColorWhite) <$ char 'N',
  (Bishop,  TeamColorWhite) <$ char 'B',
  (Rook,    TeamColorWhite) <$ char 'R',
  (Queen,   TeamColorWhite) <$ char 'Q',
  (King,    TeamColorWhite) <$ char 'K',

  (Pawn,    TeamColorBlack) <$ char 'p',
  (Knight,  TeamColorBlack) <$ char 'n',
  (Bishop,  TeamColorBlack) <$ char 'b',
  (Rook,    TeamColorBlack) <$ char 'r',
  (Queen,   TeamColorBlack) <$ char 'q',
  (King,    TeamColorBlack) <$ char 'k' ]

piecePlacement :: Parser PiecePlacement
piecePlacement = 
  (PlacementPiece <$> piece) <|>
  (PlacementInt   <$> digitToInt <$> oneOf ['1'..'8'])

piecePlacements :: Parser [[PiecePlacement]]
piecePlacements = some piecePlacement `sepBy1` char('/')

-- Active color
-- rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
--                                             ^

teamColor :: Parser TeamColor
teamColor = choice [
  TeamColorWhite <$ char 'w',
  TeamColorBlack <$ char 'b']

-- Castling availability
-- rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
--                                               ^--^
castlingAvailability :: Parser [CastlingAvailability]
castlingAvailability = 
  (pure [] <$> (char '-')) <|>
  some (
    WhiteKingside  <$ char 'K' <|>
    WhiteQueenside <$ char 'Q' <|>
    BlackKingside  <$ char 'k' <|>
    BlackQueenside <$ char 'q')

-- Enpassant Target
-- rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
--                                               ^--^

rank :: Parser Rank
rank = choice [
  RankOne    <$ char '1',
  RankTwo    <$ char '2',
  RankThree  <$ char '3',
  RankFour   <$ char '4',
  RankFive   <$ char '5',
  RankSix    <$ char '6',
  RankSeven  <$ char '7',
  RankEight  <$ char '8']

file :: Parser File
file = choice [
  FileA <$ char 'a',
  FileB <$ char 'b',
  FileC <$ char 'c',
  FileD <$ char 'd',
  FileE <$ char 'e',
  FileF <$ char 'f',
  FileG <$ char 'g',
  FileH <$ char 'h']

position :: Parser (File, Rank)
-- position = (,) <$> file <*> rank
position = liftM2 (,) file rank

enpassantTarget :: Parser (Maybe (File, Rank))
enpassantTarget = do
  (Nothing <$ (char '-')) <|> (Just <$> position)

-- Move Clocks
-- rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
--                                                      ^-^
halfmoveClock = L.decimal
fullmoveClock = L.decimal

fen :: Parser ([[PiecePlacement]], TeamColor, [CastlingAvailability], Maybe (File, Rank), Int, Int)
fen = (,,,,,) <$> 
  (piecePlacements <* space1) <*> 
  (teamColor <* space1) <*>
  (castlingAvailability <* space1) <*>
  (enpassantTarget <* space1) <*>
  (halfmoveClock <* space1) <*> fullmoveClock


-- rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
main = do
  -- input <- fmap head getArgs
  -- putStrLn (show input)
  case (parse (fen) "" "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq e3 0 1") of
    Left bundle -> putStrLn (errorBundlePretty bundle)
    Right xs -> putStrLn $ show xs
  -- parseTest (piecePlacement) "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR"

File: DanChess Shared/FENParser/FENParser.swift
//
//  FENStrings.swift
//  DanChess
//
//  Created by Daniel Beard on 6/6/21.
//  Copyright © 2021 dbeard. All rights reserved.
//

import Foundation
import SwiftParsec

/**
 A FEN "record" defines a particular game position, all in one text line and using only the ASCII character set. A text file with only FEN data records should have the file extension ".fen".[4]
 A FEN record contains six fields. The separator between fields is a space. The fields are:[5]

 1. Piece placement (from White's perspective). Each rank is described, starting with rank 8 and ending with rank 1; within each rank, the contents of each square are described from file "a" through file "h". Following the Standard Algebraic Notation (SAN), each piece is identified by a single letter taken from the standard English names (pawn = "P", knight = "N", bishop = "B", rook = "R", queen = "Q" and king = "K"). White pieces are designated using upper-case letters ("PNBRQK") while black pieces use lowercase ("pnbrqk"). Empty squares are noted using digits 1 through 8 (the number of empty squares), and "/" separates ranks.

 2. Active color. "w" means White moves next, "b" means Black moves next.

 3. Castling availability. If neither side can castle, this is "-". Otherwise, this has one or more letters: "K" (White can castle kingside), "Q" (White can castle queenside), "k" (Black can castle kingside), and/or "q" (Black can castle queenside). A move that temporarily prevents castling does not negate this notation.

 4. En passant target square in algebraic notation. If there's no en passant target square, this is "-". If a pawn has just made a two-square move, this is the position "behind" the pawn. This is recorded regardless of whether there is a pawn in position to make an en passant capture.[6]

 5. Halfmove clock: This is the number of halfmoves since the last capture or pawn advance. The reason for this field is that the value is used in the fifty-move rule.[7]

 6. Fullmove number: The number of the full move. It starts at 1, and is incremented after Black's move.

 ref (https://en.wikipedia.org/wiki/Forsyth–Edwards_Notation)
 */

func pure<A>(_ result: A) -> GenericParser<String, (), A> { GenericParser(result: result) }

public struct FENParts {
    //MARK: Raw parser values
    let piecePlacement: [[Character]]
    let activeColor: TeamColor
    let castlingAvailability: [Character]
    let enpassantTarget: Position?
    let halfmoveClock: Int?
    let fullmoveClock: Int?

    //MARK: Transformed values
    var transformedPieces: Array2D<Piece>?
    var whiteCanCastleQueenside = false
    var whiteCanCastleKingside = false
    var blackCanCastleQueenside = false
    var blackCanCastleKingside = false
}

public let FENParser: GenericParser<String, (), FENParts> = {

    // Definitions + Setup
    let fen = LanguageDefinition<()>.empty
    let lexer = GenericTokenParser(languageDefinition: fen)
    let symbol = lexer.symbol
    let char = StringParser.character

    // Piece placements
    // rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
    // ^-----------------------------------------^
    let pieces = StringParser.oneOf("RrNnBbQqKkpP")
    let emptySquares = StringParser.oneOf("1"..."8")
    let placements = (pieces <|> emptySquares).many1.separatedBy1(char("/"))
    <* lexer.whiteSpace <?> "piece placement"

    // Active color
    // rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
    //                                             ^
    let whiteColorActive = char("w") *> pure(TeamColor.white)
    let blackColorActive = char("b") *> pure(TeamColor.black)
    let activeTeamColor = (whiteColorActive <|> blackColorActive)
    <* lexer.whiteSpace <?> "team color"

    // Castling availability
    // rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
    //                                               ^--^
    let castling = (StringParser.oneOf("kqKQ").many1 <|> (char("-") *> pure(["-"])))
    <* lexer.whiteSpace <?> "castling availability"

    // Enpassant target
    // rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
    // e.g. e3                                            ^
    let rank = StringParser.oneOf("12345678")
    let file = StringParser.oneOf("abcdefgh")
    let position = file >>- { f in
        rank >>- { r in
            return pure(Position(Rank(String(r)), File(String(f))))
        }
    }
    let enpassantTarget = (position <|>
                           (char("-") *> pure(nil) ))
    <* lexer.whiteSpace.optional <?> "enpassant target"

    // Move clocks
    // rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
    //                                                      ^-^
    let halfMove = StringParser.digit.many1.stringValue.map { Int($0)! }
    <* lexer.whiteSpace.optional <?> "half move clock"

    let fullMove = StringParser.digit.many1.stringValue.map { Int($0)! }
    <?> "full move clock"

    // Parse
    return GenericParser.lift6({
        FENParts(piecePlacement: $0,
                 activeColor: $1,
                 castlingAvailability: $2,
                 enpassantTarget: $3,
                 halfmoveClock: $4,
                 fullmoveClock: $5)
    },
                               parser1: placements,
                               parser2: activeTeamColor,
                               parser3: castling,
                               parser4: enpassantTarget,
                               parser5: halfMove.optional,
                               parser6: fullMove.optional)
    }()

/// Starting position FEN
/// rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1

func fenstrings() {
    do {
        var fenParts = try FENParser.run(sourceName: "",
                                         input: "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1")
        try fenParts.transform()
        print("fenParts: \(fenParts)")
    } catch {
        print("Caught error: \(error)")
    }
}

